test/python/utils/test_lazy_loaders.py

```diff
@@ -0,0 +1,349 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Tests for the lazy loaders."""
+
+import sys
+from unittest import mock
+
+import ddt
+
+from qiskit.exceptions import MissingOptionalLibraryError
+from qiskit.test import QiskitTestCase
+from qiskit.utils import LazyImportTester, LazySubprocessTester
+
+# pylint: disable=no-member
+
+
+def available_importer(**kwargs):
+    """A LazyImportTester that should succeed."""
+    return LazyImportTester("site", **kwargs)
+
+
+def unavailable_importer(**kwargs):
+    """A LazyImportTester that should fail."""
+    return LazyImportTester("_qiskit_this_module_does_not_exist_", **kwargs)
+
+
+def available_process(**kwargs):
+    """A LazySubprocessTester that should fail."""
+    return LazySubprocessTester([sys.executable, "-c", "import sys; sys.exit(0)"], **kwargs)
+
+
+def unavailable_process(**kwargs):
+    """A LazySubprocessTester that should fail."""
+    return LazySubprocessTester([sys.executable, "-c", "import sys; sys.exit(1)"], **kwargs)
+
+
+def mock_availability_test(feature):
+    """Context manager that mocks out the availability checker for a given dependency checker.  The
+    context manager returns the mocked-out method."""
+    # We have to be careful with what we patch because the dependency managers define `__slots__`.
+    return mock.patch.object(type(feature), "_is_available", wraps=feature._is_available)
+
+
+@ddt.ddt
+class TestLazyDependencyTester(QiskitTestCase):
+    """Tests for the lazy loaders.  Within this class, we parameterise the test cases with
+    generators, rather than the mocks themselves.  That allows us to easily generate clean
+    instances, and means that creation doesn't happen during test collection."""
+
+    @ddt.data(available_importer, available_process)
+    def test_evaluates_correctly_true(self, test_generator):
+        """Test that the available loaders evaluate True in various Boolean contexts."""
+        self.assertTrue(test_generator())
+        self.assertTrue(bool(test_generator()))
+        if not test_generator():
+            self.fail("did not evaluate true")
+
+    @ddt.data(unavailable_importer, unavailable_process)
+    def test_evaluates_correctly_false(self, test_generator):
+        """Test that the available loaders evaluate False in various Boolean contexts."""
+        self.assertFalse(test_generator())
+        self.assertFalse(bool(test_generator()))
+        if test_generator():
+            self.fail("did not evaluate false")
```

Comment:
Is there an advantage to this vs L71? This should be equivalent to `self.assertFalse(test_generator())`

Reply:
Yeah, it's a slightly superfluous test - I'm not 100% sure what I was thinking, but I suspect it was along the lines of "check a few possible implicit Boolean contexts".  Strictly, all three tests here should be identical, I think.  Happy to remove it (and the similar test on line 65 if you'd prefer to reduce noise.